### ISSUE TITLE
Pattern match support

### ISSUE THREAD

**Issue by sergey-lukianov** [kuon/ex-pipe-helpers#2]
It would be great to add pattern match support into `tap_on` and `then_on`

For example
```
%{a: 1, b:2}
|> tap_on(%{a:1}, fn _-> IO.puts("Win!") end)
```

**Comment by kuon** [comment]
Yes, this would be a nice touch, I thought of it, but didn't find a way to leverage built in pattern matching for that.

It would be possible to just do:

```elixir

def tap_on(%{} = actual, %{} = ref, fun) do
  if Map.get(actual, Map.keys(ref)) |> Map.equal?(ref) do
 ...
```

But that wouldn't really be similar to the real elixir pattern matching. And there might be some gotcha I am not thinking of right now.

The other solution would be to write a macro that generates the following code:

```elixir
%{a: 1, b: 2}
|> case do
  %{a: 1} = val -> fun(val); val 
  val -> val
end
```

I think this would be better.

But also here, maybe some gotcha I am not thinking of right now.
